File: python/grafana_utils/clients/alert_client.py
"""Alerting-focused Grafana API client helpers."""

from typing import Any, Optional
from urllib import parse

from ..alerts.common import GrafanaApiError, GrafanaError
from ..http_transport import (
    HttpTransportApiError,
    HttpTransportError,
    JsonHttpTransport,
    build_json_http_transport,
)


class GrafanaAlertClient:
    """Minimal HTTP wrapper around the Grafana alerting provisioning APIs."""

    def __init__(
        self,
        base_url: str,
        headers: dict[str, str],
        timeout: int,
        verify_ssl: bool,
        transport: Optional[JsonHttpTransport] = None,
    ) -> None:
        # Purpose: implementation note.
        # Args: see function signature.
        # Returns: see implementation.

        # Call graph: see callers/callees.
        #   Upstream callers: 無
        #   Downstream callees: 無

        self.transport = transport or build_json_http_transport(
            base_url=base_url,
            headers={"Accept": "application/json", **headers},
            timeout=timeout,
            verify_ssl=verify_ssl,
        )

    def request_json(
        self,
        path: str,
        params: Optional[dict[str, Any]] = None,
        method: str = "GET",
        payload: Optional[dict[str, Any]] = None,
    ) -> Any:
        """Send one request to Grafana and decode the JSON response body."""
        try:
            return self.transport.request_json(
                path=path,
                params=params,
                method=method,
                payload=payload,
            )
        except HttpTransportApiError as exc:
            raise GrafanaApiError(exc.status_code, exc.url, exc.body) from exc
        except HttpTransportError as exc:
            raise GrafanaError(str(exc)) from exc

    def list_alert_rules(self) -> list[dict[str, Any]]:
        # Purpose: implementation note.
        # Args: see function signature.
        # Returns: see implementation.

        # Call graph: see callers/callees.
        #   Upstream callers: 無
        #   Downstream callees: 37

        data = self.request_json("/api/v1/provisioning/alert-rules")
        if not isinstance(data, list):
            raise GrafanaError("Unexpected alert-rule list response from Grafana.")
        return [item for item in data if isinstance(item, dict)]

    def search_dashboards(self, query: str) -> list[dict[str, Any]]:
        # Purpose: implementation note.
        # Args: see function signature.
        # Returns: see implementation.

        # Call graph: see callers/callees.
        #   Upstream callers: 無
        #   Downstream callees: 37

        data = self.request_json(
            "/api/search",
            params={"type": "dash-db", "query": query, "limit": 500},
        )
        if not isinstance(data, list):
            raise GrafanaError("Unexpected dashboard search response from Grafana.")
        return [item for item in data if isinstance(item, dict)]

    def get_dashboard(self, uid: str) -> dict[str, Any]:
        # Purpose: implementation note.
        # Args: see function signature.
        # Returns: see implementation.

        # Call graph: see callers/callees.
        #   Upstream callers: 無
        #   Downstream callees: 37

        data = self.request_json("/api/dashboards/uid/%s" % parse.quote(uid, safe=""))
        if not isinstance(data, dict):
            raise GrafanaError("Unexpected dashboard payload for UID %s." % uid)
        return data

    def get_alert_rule(self, uid: str) -> dict[str, Any]:
        # Purpose: implementation note.
        # Args: see function signature.
        # Returns: see implementation.

        # Call graph: see callers/callees.
        #   Upstream callers: 無
        #   Downstream callees: 37

        data = self.request_json(
            "/api/v1/provisioning/alert-rules/%s" % parse.quote(uid, safe="")
        )
        if not isinstance(data, dict):
            raise GrafanaError("Unexpected alert-rule payload for UID %s." % uid)
        return data

    def create_alert_rule(self, payload: dict[str, Any]) -> dict[str, Any]:
        # Purpose: implementation note.
        # Args: see function signature.
        # Returns: see implementation.

        # Call graph: see callers/callees.
        #   Upstream callers: 無
        #   Downstream callees: 37

        data = self.request_json(
            "/api/v1/provisioning/alert-rules",
            method="POST",
            payload=payload,
        )
        if not isinstance(data, dict):
            raise GrafanaError("Unexpected alert-rule create response from Grafana.")
        return data

    def update_alert_rule(self, uid: str, payload: dict[str, Any]) -> dict[str, Any]:
        # Purpose: implementation note.
        # Args: see function signature.
        # Returns: see implementation.

        # Call graph: see callers/callees.
        #   Upstream callers: 無
        #   Downstream callees: 37

        data = self.request_json(
            "/api/v1/provisioning/alert-rules/%s" % parse.quote(uid, safe=""),
            method="PUT",
            payload=payload,
        )
        if not isinstance(data, dict):
            raise GrafanaError("Unexpected alert-rule update response from Grafana.")
        return data

    def list_contact_points(self) -> list[dict[str, Any]]:
        # Purpose: implementation note.
        # Args: see function signature.
        # Returns: see implementation.

        # Call graph: see callers/callees.
        #   Upstream callers: 無
        #   Downstream callees: 37

        data = self.request_json("/api/v1/provisioning/contact-points")
        if not isinstance(data, list):
            raise GrafanaError("Unexpected contact-point list response from Grafana.")
        return [item for item in data if isinstance(item, dict)]

    def create_contact_point(self, payload: dict[str, Any]) -> dict[str, Any]:
        # Purpose: implementation note.
        # Args: see function signature.
        # Returns: see implementation.

        # Call graph: see callers/callees.
        #   Upstream callers: 無
        #   Downstream callees: 37

        data = self.request_json(
            "/api/v1/provisioning/contact-points",
            method="POST",
            payload=payload,
        )
        if not isinstance(data, dict):
            raise GrafanaError("Unexpected contact-point create response from Grafana.")
        return data

    def update_contact_point(self, uid: str, payload: dict[str, Any]) -> dict[str, Any]:
        # Purpose: implementation note.
        # Args: see function signature.
        # Returns: see implementation.

        # Call graph: see callers/callees.
        #   Upstream callers: 無
        #   Downstream callees: 37

        data = self.request_json(
            "/api/v1/provisioning/contact-points/%s" % parse.quote(uid, safe=""),
            method="PUT",
            payload=payload,
        )
        if not isinstance(data, dict):
            raise GrafanaError("Unexpected contact-point update response from Grafana.")
        return data

    def list_mute_timings(self) -> list[dict[str, Any]]:
        # Purpose: implementation note.
        # Args: see function signature.
        # Returns: see implementation.

        # Call graph: see callers/callees.
        #   Upstream callers: 無
        #   Downstream callees: 37

        data = self.request_json("/api/v1/provisioning/mute-timings")
        if not isinstance(data, list):
            raise GrafanaError("Unexpected mute-timing list response from Grafana.")
        return [item for item in data if isinstance(item, dict)]

    def create_mute_timing(self, payload: dict[str, Any]) -> dict[str, Any]:
        # Purpose: implementation note.
        # Args: see function signature.
        # Returns: see implementation.

        # Call graph: see callers/callees.
        #   Upstream callers: 無
        #   Downstream callees: 37

        data = self.request_json(
            "/api/v1/provisioning/mute-timings",
            method="POST",
            payload=payload,
        )
        if not isinstance(data, dict):
            raise GrafanaError("Unexpected mute-timing create response from Grafana.")
        return data

    def update_mute_timing(self, name: str, payload: dict[str, Any]) -> dict[str, Any]:
        # Purpose: implementation note.
        # Args: see function signature.
        # Returns: see implementation.

        # Call graph: see callers/callees.
        #   Upstream callers: 無
        #   Downstream callees: 37

        data = self.request_json(
            "/api/v1/provisioning/mute-timings/%s" % parse.quote(name, safe=""),
            method="PUT",
            payload=payload,
        )
        if not isinstance(data, dict):
            raise GrafanaError("Unexpected mute-timing update response from Grafana.")
        return data

    def get_notification_policies(self) -> dict[str, Any]:
        # Purpose: implementation note.
        # Args: see function signature.
        # Returns: see implementation.

        # Call graph: see callers/callees.
        #   Upstream callers: 無
        #   Downstream callees: 37

        data = self.request_json("/api/v1/provisioning/policies")
        if not isinstance(data, dict):
            raise GrafanaError("Unexpected notification policy response from Grafana.")
        return data

    def update_notification_policies(self, payload: dict[str, Any]) -> dict[str, Any]:
        # Purpose: implementation note.
        # Args: see function signature.
        # Returns: see implementation.

        # Call graph: see callers/callees.
        #   Upstream callers: 無
        #   Downstream callees: 37

        data = self.request_json(
            "/api/v1/provisioning/policies",
            method="PUT",
            payload=payload,
        )
        if not isinstance(data, dict):
            raise GrafanaError(
                "Unexpected notification policy update response from Grafana."
            )
        return data

    def list_templates(self) -> list[dict[str, Any]]:
        # Purpose: implementation note.
        # Args: see function signature.
        # Returns: see implementation.

        # Call graph: see callers/callees.
        #   Upstream callers: 無
        #   Downstream callees: 37

        data = self.request_json("/api/v1/provisioning/templates")
        if data is None:
            return []
        if not isinstance(data, list):
            raise GrafanaError("Unexpected template list response from Grafana.")
        return [item for item in data if isinstance(item, dict)]

    def get_template(self, name: str) -> dict[str, Any]:
        # Purpose: implementation note.
        # Args: see function signature.
        # Returns: see implementation.

        # Call graph: see callers/callees.
        #   Upstream callers: 無
        #   Downstream callees: 37

        data = self.request_json(
            "/api/v1/provisioning/templates/%s" % parse.quote(name, safe="")
        )
        if not isinstance(data, dict):
            raise GrafanaError("Unexpected template payload for name %s." % name)
        return data

    def update_template(self, name: str, payload: dict[str, Any]) -> dict[str, Any]:
        # Purpose: implementation note.
        # Args: see function signature.
        # Returns: see implementation.

        # Call graph: see callers/callees.
        #   Upstream callers: 無
        #   Downstream callees: 37

        body = dict(payload)
        body.pop("name", None)
        data = self.request_json(
            "/api/v1/provisioning/templates/%s" % parse.quote(name, safe=""),
            method="PUT",
            payload=body,
        )
        if not isinstance(data, dict):
            raise GrafanaError("Unexpected template update response from Grafana.")
        return data

File: rust/src/alert_client.rs
//! Minimal Grafana alerting API client.
//! Wraps typed request wrappers for list/import/export flows and hides raw HTTP plumbing from CLI handlers.
use reqwest::Method;
use serde_json::{Map, Value};

use crate::common::{message, Result};
use crate::http::{JsonHttpClient, JsonHttpClientConfig};

use super::AlertAuthContext;

pub struct GrafanaAlertClient {
    http: JsonHttpClient,
}

impl GrafanaAlertClient {
    pub fn new(context: &AlertAuthContext) -> Result<Self> {
        Ok(Self {
            http: JsonHttpClient::new(JsonHttpClientConfig {
                base_url: context.url.clone(),
                headers: context.headers.clone(),
                timeout_secs: context.timeout,
                verify_ssl: context.verify_ssl,
            })?,
        })
    }

    fn request_json(
        &self,
        method: Method,
        path: &str,
        params: &[(String, String)],
        payload: Option<&Value>,
    ) -> Result<Option<Value>> {
        self.http.request_json(method, path, params, payload)
    }

    pub fn list_alert_rules(&self) -> Result<Vec<Map<String, Value>>> {
        expect_object_list(
            self.request_json(Method::GET, "/api/v1/provisioning/alert-rules", &[], None)?,
            "Unexpected alert-rule list response from Grafana.",
        )
    }

    pub fn search_dashboards(&self, query: &str) -> Result<Vec<Map<String, Value>>> {
        expect_object_list(
            self.request_json(
                Method::GET,
                "/api/search",
                &[
                    ("type".to_string(), "dash-db".to_string()),
                    ("query".to_string(), query.to_string()),
                    ("limit".to_string(), "500".to_string()),
                ],
                None,
            )?,
            "Unexpected dashboard search response from Grafana.",
        )
    }

    pub fn get_dashboard(&self, uid: &str) -> Result<Map<String, Value>> {
        expect_object(
            self.request_json(
                Method::GET,
                &format!("/api/dashboards/uid/{uid}"),
                &[],
                None,
            )?,
            &format!("Unexpected dashboard payload for UID {uid}."),
        )
    }

    pub fn get_alert_rule(&self, uid: &str) -> Result<Map<String, Value>> {
        expect_object(
            self.request_json(
                Method::GET,
                &format!("/api/v1/provisioning/alert-rules/{uid}"),
                &[],
                None,
            )?,
            &format!("Unexpected alert-rule payload for UID {uid}."),
        )
    }

    pub fn create_alert_rule(&self, payload: &Map<String, Value>) -> Result<Map<String, Value>> {
        expect_object(
            self.request_json(
                Method::POST,
                "/api/v1/provisioning/alert-rules",
                &[],
                Some(&Value::Object(payload.clone())),
            )?,
            "Unexpected alert-rule create response from Grafana.",
        )
    }

    pub fn update_alert_rule(
        &self,
        uid: &str,
        payload: &Map<String, Value>,
    ) -> Result<Map<String, Value>> {
        expect_object(
            self.request_json(
                Method::PUT,
                &format!("/api/v1/provisioning/alert-rules/{uid}"),
                &[],
                Some(&Value::Object(payload.clone())),
            )?,
            "Unexpected alert-rule update response from Grafana.",
        )
    }

    pub fn list_contact_points(&self) -> Result<Vec<Map<String, Value>>> {
        expect_object_list(
            self.request_json(
                Method::GET,
                "/api/v1/provisioning/contact-points",
                &[],
                None,
            )?,
            "Unexpected contact-point list response from Grafana.",
        )
    }

    pub fn create_contact_point(&self, payload: &Map<String, Value>) -> Result<Map<String, Value>> {
        expect_object(
            self.request_json(
                Method::POST,
                "/api/v1/provisioning/contact-points",
                &[],
                Some(&Value::Object(payload.clone())),
            )?,
            "Unexpected contact-point create response from Grafana.",
        )
    }

    pub fn update_contact_point(
        &self,
        uid: &str,
        payload: &Map<String, Value>,
    ) -> Result<Map<String, Value>> {
        expect_object(
            self.request_json(
                Method::PUT,
                &format!("/api/v1/provisioning/contact-points/{uid}"),
                &[],
                Some(&Value::Object(payload.clone())),
            )?,
            "Unexpected contact-point update response from Grafana.",
        )
    }

    pub fn list_mute_timings(&self) -> Result<Vec<Map<String, Value>>> {
        expect_object_list(
            self.request_json(Method::GET, "/api/v1/provisioning/mute-timings", &[], None)?,
            "Unexpected mute-timing list response from Grafana.",
        )
    }

    pub fn create_mute_timing(&self, payload: &Map<String, Value>) -> Result<Map<String, Value>> {
        expect_object(
            self.request_json(
                Method::POST,
                "/api/v1/provisioning/mute-timings",
                &[],
                Some(&Value::Object(payload.clone())),
            )?,
            "Unexpected mute-timing create response from Grafana.",
        )
    }

    pub fn update_mute_timing(
        &self,
        name: &str,
        payload: &Map<String, Value>,
    ) -> Result<Map<String, Value>> {
        expect_object(
            self.request_json(
                Method::PUT,
                &format!("/api/v1/provisioning/mute-timings/{name}"),
                &[],
                Some(&Value::Object(payload.clone())),
            )?,
            "Unexpected mute-timing update response from Grafana.",
        )
    }

    pub fn get_notification_policies(&self) -> Result<Map<String, Value>> {
        expect_object(
            self.request_json(Method::GET, "/api/v1/provisioning/policies", &[], None)?,
            "Unexpected notification policy response from Grafana.",
        )
    }

    pub fn update_notification_policies(
        &self,
        payload: &Map<String, Value>,
    ) -> Result<Map<String, Value>> {
        expect_object(
            self.request_json(
                Method::PUT,
                "/api/v1/provisioning/policies",
                &[],
                Some(&Value::Object(payload.clone())),
            )?,
            "Unexpected notification policy update response from Grafana.",
        )
    }

    pub fn list_templates(&self) -> Result<Vec<Map<String, Value>>> {
        parse_template_list_response(self.request_json(
            Method::GET,
            "/api/v1/provisioning/templates",
            &[],
            None,
        )?)
    }

    pub fn get_template(&self, name: &str) -> Result<Map<String, Value>> {
        expect_object(
            self.request_json(
                Method::GET,
                &format!("/api/v1/provisioning/templates/{name}"),
                &[],
                None,
            )?,
            &format!("Unexpected template payload for name {name}."),
        )
    }

    pub fn update_template(
        &self,
        name: &str,
        payload: &Map<String, Value>,
    ) -> Result<Map<String, Value>> {
        let mut body = payload.clone();
        body.remove("name");
        expect_object(
            self.request_json(
                Method::PUT,
                &format!("/api/v1/provisioning/templates/{name}"),
                &[],
                Some(&Value::Object(body)),
            )?,
            "Unexpected template update response from Grafana.",
        )
    }
}

fn expect_object(value: Option<Value>, error_message: &str) -> Result<Map<String, Value>> {
    match value {
        Some(Value::Object(object)) => Ok(object),
        _ => Err(message(error_message)),
    }
}

pub fn expect_object_list(
    value: Option<Value>,
    error_message: &str,
) -> Result<Vec<Map<String, Value>>> {
    let array = match value {
        Some(Value::Array(items)) => items,
        _ => return Err(message(error_message)),
    };
    Ok(array
        .into_iter()
        .filter_map(|item| match item {
            Value::Object(object) => Some(object),
            _ => None,
        })
        .collect())
}

// Interpret Grafana template-list responses in the one supported shape for
// alerting imports: either empty/null or a JSON array of template objects.
pub fn parse_template_list_response(value: Option<Value>) -> Result<Vec<Map<String, Value>>> {
    match value {
        None | Some(Value::Null) => Ok(Vec::new()),
        Some(value) => expect_object_list(
            Some(value),
            "Unexpected template list response from Grafana.",
        ),
    }
}
